### PR TITLE
[TEP-0091] add v1 support for VerifyResource

### PR DESCRIPTION
**Pull request by Yongxuanzhang**
# Changes
This commit adds v1 task and pipeline verification for VerifyResource. This is part of the v1 support for trusted resources.

/kind feature

Part of #6356

Signed-off-by: Yongxuan Zhang yongxuanzhang@google.com

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [ ] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
NONE
```

### CODE REVIEW

**Review by dibyom — APPROVED** [review]

File: pkg/trustedresources/verify.go [hunk 1]
```diff
@@ -63,14 +64,13 @@ type VerificationResult struct {
...
+// VerifyResource verifies the signature and public key against resource (v1beta1 and v1 task and pipeline).
 // VerificationResult is returned with different types for different cases:
 // 1) Return VerificationResult with VerificationSkip type, when no policies are found and no-match-policy is set to ignore
 // 2) Return VerificationResult with VerificationPass type when verification passed;
 // 3) Return VerificationResult with VerificationWarn type, when no matching policies and feature flag "no-match-policy" is "warn", or only Warn mode verification policies fail. Err field is filled with the warning;
 // 4) Return VerificationResult with VerificationError type when no policies are found and no-match-policy is set to fail, the resource fails to pass matched enforce verification policy, or there are errors during verification. Err is filled with the err.
 // refSource contains the source information of the resource.
-// TODO(#6729): Support v1 task and pipeline
 func VerifyResource(ctx context.Context, resource metav1.Object, k8s kubernetes.Interface, refSource *v1beta1.RefSource, verificationpolicies []*v1alpha1.VerificationPolicy) VerificationResult {
```
[the remark this answers is not in the document]
Comment (Yongxuanzhang) [reply pkg/trustedresources/verify.go:74]: Oh, I thought you're asking resource version. 
This needs to be updated to v1.RefSource. It is ok to update it in the swap PR

**Review by JeromeJu — APPROVED** [review]
Thanks @Yongxuanzhang , one remaining question I had is how are we going to deal with verification of v1.PipelineRef and v1.TaskRef? and how these testing should be covered when we swap to v1 with the v1 verification support?

**Comment by JeromeJu** [comment]
/lgtm

Thanks @Yongxuanzhang , one remaining question I had is how are we going to deal with verification of v1.PipelineRef and v1.TaskRef? and how these testing should be covered when we swap to v1 with the v1 verification support?

There will be 2 followup PRs to add v1 support in taskref.go and pipelineref.go, and also add tests there. If this is what you're asking?

also SGTM

**Comment by Yongxuanzhang** [comment]
I don't know what's going on with the `pull-tekton-pipeline-go-coverage-df` test ...

**Comment by lbernick** [comment]
> I don't know what's going on with the `pull-tekton-pipeline-go-coverage-df` test ...

I filed https://github.com/tektoncd/plumbing/issues/1460 to track this

**Comment by Yongxuanzhang** [comment]
> tektoncd/plumbing#1460

Oh thanks! I think it's something worth looking into. 🤔 We have 2 coverage jobs in ci. I think the goal of having the -df one is for dogfooding, there might be some issues inside with it...

**Comment by Yongxuanzhang** [comment]
/retest 
https://github.com/tektoncd/pipeline/issues/6730
log: https://prow.tekton.dev/view/gs/tekton-prow/pr-logs/pull/tektoncd_pipeline/6738/pull-tekton-pipeline-integration-tests/1664288989609201664

File: pkg/trustedresources/verify.go [hunk 2]
```diff
@@ -120,6 +133,19 @@ func VerifyResource(ctx context.Context, resource metav1.Object, k8s kubernetes.
...
+		pipeline := v1.Pipeline{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "tekton.dev/v1",
+				Kind:       "Pipeline"},
+			ObjectMeta: pm,
+			Spec:       v.Spec,
+		}
+		return verifyResource(ctx, &pipeline, k8s, signature, matchedPolicies)
 	}
 	return VerificationResult{VerificationResultType: VerificationError, Err: fmt.Errorf("%w: got resource %v but v1beta1.Task and v1beta1.Pipeline are currently supported", ErrResourceNotSupported, resource)}
```
Comment (chuangw6) [inline pkg/trustedresources/verify.go:150]: move this line to the default case of the switch statement?

File: pkg/trustedresources/verify.go [hunk 3]
```diff
@@ -107,6 +107,19 @@ func VerifyResource(ctx context.Context, resource metav1.Object, k8s kubernetes.
 			Spec:       v.TaskSpec(),
 		}
 		return verifyResource(ctx, &task, k8s, signature, matchedPolicies)
+	case *v1.Task:
+		tm, signature, err := prepareObjectMeta(v.ObjectMeta)
+		if err != nil {
+			return VerificationResult{VerificationResultType: VerificationError, Err: err}
+		}
```
Comment (chuangw6) [inline pkg/trustedresources/verify.go:114]: This code chunk repeats in all 4 switch cases. We can change the function `prepareObjectMeta` 

from accepting struct
```go
func prepareObjectMeta(in metav1.ObjectMeta) (metav1.ObjectMeta, []byte, error) {
```
to accepting interface
```go
func prepareObjectMeta(in metav1.Object) (metav1.ObjectMeta, []byte, error) {
```

This way, we can avoid duplicate code by moving this out of switch statement.

Comment (Yongxuanzhang) [reply pkg/trustedresources/verify.go]: Oh thanks! Great suggestion

**Pull request merged**